File: js_port/fs_js/src/model/DelegateFunction.js
const { BaseFunction, CallType } = require('../core/functionBase');
const { ensureTyped } = require('../core/value');

class DelegateFunction extends BaseFunction {
  constructor(delegate) {
    super();
    if (typeof delegate !== 'function') {
      throw new Error('DelegateFunction requires a function');
    }
    this.delegate = delegate;
  }

  get maxParameters() {
    return -1;
  }

  evaluate(provider, parameters) {
    const args = [];
    const count = parameters ? parameters.count : 0;
    for (let i = 0; i < count; i += 1) {
      args.push(parameters.getParameter(provider, i));
    }
    const result = this.delegate(...args.map(ensureTyped));
    return ensureTyped(result);
  }

  get callType() {
    return CallType.Infix;
  }
}

module.exports = {
  DelegateFunction
};

File: FuncScript/Model/DelegateFunction.cs
using FuncScript.Core;
using FuncScript.Functions.Logic;
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace FuncScript.Model
{
    internal class ByteArray 
    {
        public byte[] Bytes;
    }

    internal class DelegateFunction : IFsFunction
    {
        private Delegate f;
        System.Reflection.ParameterInfo[] _pars;
        public DelegateFunction(Delegate f)
        {
            this.f = f;
            var m = f.Method;
            if (m.ReturnType == typeof(void))
                throw new Error.TypeMismatchError("Delegate with no return is not supported");
            _pars = m.GetParameters();
            foreach(var p in _pars)
            {
                if(p.IsOut)
                    throw new Error.TypeMismatchError($"Delegate with output parameters not supported. Par:{p.Name}");
            }
        }

        public int MaxParsCount => _pars.Length;

        public CallType CallType => CallType.Infix;

        public string Symbol => throw new NotSupportedException();

        public int Precedence => 0;

        public object Evaluate(IFsDataProvider parent, IParameterList pars)
        {
            return this.f.DynamicInvoke(Enumerable.Range(0, pars.Count).Select(x => pars.GetParameter(parent, x)).ToArray());
        }

        public string ParName(int index)
        {
            throw new NotImplementedException();
        }
    }
}
